source/material_model/grain_size.cc

```diff
@@ -222,165 +241,187 @@ namespace aspect
-      if ((original_grain_size != original_grain_size) || this->get_timestep() == 0.0
-          || original_grain_size < std::numeric_limits<double>::min())
-        return 0.0;
-
-      // set up the parameters for the sub-timestepping of grain size evolution
-      std::vector<double> current_composition = compositional_fields;
-      double grain_size = original_grain_size;
-      double grain_size_change = 0.0;
-      const double timestep = this->get_timestep();
-
-      // use a sub timestep of 500 yrs, currently fixed timestep
-      double grain_growth_timestep = 500 * 3600 * 24 * 365.25;
-      double time = 0;
-
-      // precompute the partitioning_fraction since its constant during the evolution.
-      // this is only used for the pinned_grain_damage formulation
-      const double partitioning_fraction = (grain_size_evolution_formulation == Formulation::pinned_grain_damage)
-                                           ?
-                                           compute_partitioning_fraction(temperature)
-                                           :
-                                           0.0;
-
-      // we keep the dislocation viscosity of the last iteration as guess
-      // for the next one
-      double current_dislocation_viscosity = 0.0;
-
-      const double adiabatic_temperature = this->get_adiabatic_conditions().is_initialized()
-                                           ?
-                                           this->get_adiabatic_conditions().temperature(position)
-                                           :
-                                           temperature;
-
-      do
+      std::vector<double> grain_size_changes (in.n_evaluation_points());
+      std::vector<std::vector<double>> reaction_terms (in.n_evaluation_points(), std::vector<double>(this->n_compositional_fields(), 0.0));
+
+      // Set up an integer that tells us which phase transition has been crossed inside of the cell.
+      std::vector<int> crossed_transitions (in.n_evaluation_points(), -1);
+
+      grain_sizes.reinit(in.n_evaluation_points());
+
+      const double timestep = this->simulator_is_past_initialization()
+                              ?
+                              this->get_timestep()
+                              :
+                              0.0;
+
+      for (unsigned int i=0; i<in.n_evaluation_points(); ++i)
         {
-          time += grain_growth_timestep;
+          grain_sizes[i]=in.composition[i][grain_size_index];
 
-          if (timestep - time < 0)
-            {
-              grain_growth_timestep = timestep - (time - grain_growth_timestep);
-              time = timestep;
-            }
+          if ((in.composition[i][grain_size_index] != in.composition[i][grain_size_index]) || this->get_timestep() == 0.0
+              || in.composition[i][grain_size_index] < std::numeric_limits<double>::min())
+            return reaction_terms;
```

Comment:
I don't understand this. The function is supposed to return grain size changes (based on its name) but returns reaction rates here. This cannot be correct based on units alone. A grain size change has units meters, the reation rate has units meters per second.

Separately, did you mean to exit out of the function if *any* of the points satisfy this condition? What happens to the other points?

Reply:
The reaction terms are not reaction rates, they are changes in the compositional fields. That is the variable name we use in all material models.